PSWebServiceLibrary.php

```diff
@@ -92,50 +90,66 @@ function __construct($url, $key, $debug = true)
+                throw PrestaShopWebserviceStatusException::shouldNotReceive304NotModifiedStatus();
+            case 307:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive307TemporaryRedirectStatus();
+            case 308:
+                throw PrestaShopWebserviceStatusException::shouldNotReceive308PermanentRedirectStatus();
             case 400:
-                $error_message = 'Bad Request';
-                break;
+                throw new PrestaShopWebserviceBadRequestException();
             case 401:
-                $error_message = 'Unauthorized';
-                break;
+                throw new PrestaShopWebserviceUnauthorizedException();
+            case 403:
+                throw new PrestaShopWebserviceForbiddenException();
             case 404:
-                $error_message = 'Not Found';
-                break;
+                throw new PrestaShopWebserviceNotFoundException();
             case 405:
-                $error_message = 'Method Not Allowed';
-                break;
-            case 500:
-                $error_message = 'Internal Server Error';
-                break;
-            default:
-                throw new PrestaShopWebserviceException(
-                    'This call to PrestaShop Web Services returned an unexpected HTTP status of:' . $request['status_code']
-                );
+                throw new PrestaShopWebserviceMethodNotAllowedException();
+            case 429:
+                throw new PrestaShopWebserviceTooManyRequestsException();
+        }
+        if ($statusCode >= 100 && $statusCode < 200
+            || $statusCode >= 202 && $statusCode < 300
+            || $statusCode >= 300 && $statusCode < 500
+            || $statusCode >= 600
+            || $statusCode < 100
+        ) {
+            throw new PrestaShopWebserviceClientException('This call to PrestaShop Web Services responded with a non-standard code or a code this client could not handle properly. This can come from your web server or reverse proxy configuration.', $statusCode);
         }
 
-        if (!empty($error_message)) {
-            $response = $this->parseXML($request['response']);
-            $errors = $response->children()->children();
-            if ($errors && count($errors) > 0) {
-                foreach ($errors as $error) {
-                    $error_message.= ' - (Code ' . $error->code . '): ' . $error->message;
-                }
-            }
-            $error_label = 'This call to PrestaShop Web Services failed and returned an HTTP status of %d. That means: %s.';
-            throw new PrestaShopWebserviceException(sprintf($error_label, $request['status_code'], $error_message));
+        if ($statusCode >= 500 && $statusCode < 600) {
+            throw new PrestaShopWebserviceServerException('This call to PrestaShop Web Services responded with a status code indicating it is not available or under a too heavy load to process your request.', $statusCode);
```

Comment:
You'd want this block before the previous one:

```php
if ($statusCode >= 500 && $statusCode < 600) {
    throw new PrestaShopWebserviceServerException('This call to PrestaShop Web Services responded with a status code indicating it is not available or under a too heavy load to process your request.', $statusCode);
}
throw new PrestaShopWebserviceClientException('This call to PrestaShop Web Services responded with a non-standard code or a code this client could not handle properly. This can come from your web server or reverse proxy configuration.', $statusCode);
```

That said, I think having a dedicated error for HTTP 500 would be a good idea since this is the most common error you can have if the api crashes.

Reply:
Additionally, with the example code you are providing a `PrestaShopWebserviceClientException` will be thrown on 200 and 201 response codes.